sdk/core/azure-core/test/ut/websocket_test.cpp

```diff
@@ -0,0 +1,686 @@
+// SPDX-License-Identifier: MIT
+
+#include "../../src/http/websockets/websocketsimpl.hpp"
+#include "azure/core/http/websockets/websockets.hpp"
+#include "azure/core/internal/json/json.hpp"
+#include <chrono>
+#include <gtest/gtest.h>
+#include <list>
+#include <set>
+#include <thread>
+#if defined(BUILD_CURL_HTTP_TRANSPORT_ADAPTER)
+#include "azure/core/http/websockets/curl_websockets_transport.hpp"
+#endif
+// cspell::words closeme flibbityflobbidy
+
+using namespace Azure::Core;
+using namespace Azure::Core::Http::WebSockets;
+using namespace std::chrono_literals;
+
+class WebSocketTests : public testing::Test {
+private:
+protected:
+  // Create
+  static void SetUpTestSuite() {}
+  static void TearDownTestSuite()
+  {
+    //    GTEST_LOG_(INFO) << "Shut down test server" << std::endl;
+    //    WebSocket controlSocket(Azure::Core::Url("http://localhost:8000/control"));
+    //    controlSocket.Open();
+    //    controlSocket.SendFrame("close", true);
+    //    auto controlResponse = controlSocket.ReceiveFrame();
+    //    EXPECT_EQ(controlResponse->FrameType, WebSocketFrameType::TextFrameReceived);
+  }
+};
+
+TEST_F(WebSocketTests, CreateSimpleSocket)
+{
+  {
+    WebSocket defaultSocket(Azure::Core::Url("http://localhost:8000"));
+    defaultSocket.AddHeader("newHeader", "headerValue");
+    EXPECT_THROW(defaultSocket.GetChosenProtocol(), std::runtime_error);
+  }
+}
+
+TEST_F(WebSocketTests, OpenSimpleSocket)
+{
+  {
+    WebSocketOptions options;
+    WebSocket defaultSocket(Azure::Core::Url("http://localhost:8000/openclosetest"), options);
+    defaultSocket.AddHeader("newHeader", "headerValue");
+
+    defaultSocket.Open();
+
+    EXPECT_THROW(defaultSocket.AddHeader("newHeader", "headerValue"), std::runtime_error);
+
+    // Close the socket without notifying the peer.
+    defaultSocket.Close();
+  }
+
+  {
+    WebSocketOptions options;
+    WebSocket defaultSocket(Azure::Core::Url("http://microsoft.com/index.htm"), options);
+    defaultSocket.AddHeader("newHeader", "headerValue");
+
+    // When running this test locally, the call times out, so drop in a 5 second timeout on
+    // the request.
+    Azure::Core::Context requestContext = Azure::Core::Context::ApplicationContext.WithDeadline(
+        std::chrono::system_clock::now() + 15s);
+    EXPECT_THROW(defaultSocket.Open(requestContext), std::runtime_error);
+  }
+}
+
+TEST_F(WebSocketTests, OpenAndCloseSocket)
+{
+  if (false)
+  {
+    WebSocket defaultSocket(Azure::Core::Url("http://localhost:8000/openclosetest"));
+    defaultSocket.AddHeader("newHeader", "headerValue");
+
+    defaultSocket.Open();
+
+    // Close the socket without notifying the peer.
+    defaultSocket.Close(4500);
+  }
+
+  {
+    WebSocket defaultSocket(Azure::Core::Url("http://localhost:8000/openclosetest"));
+
+    defaultSocket.Open();
+
+    // Close the socket without notifying the peer.
+    defaultSocket.Close(4500, "This is a good reason.");
```

Comment:
We use 4500 in a number of tests.  Should make it a constant

Reply:
It's twice, but I've added a manifest constant: `UndefinedButLegalCloseReason` (since that is what 4500 is: A legal close reason but not one which is defined).